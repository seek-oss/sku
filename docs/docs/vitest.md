# Vitest

[Vitest] is a testing framework that supports ESM out-of-the-box, integrates with the Vite ecosystem and has a similar API to Jest.
These features make it a great replacement for Jest in `sku` applications, especially given [Jest's current limitations with ESM].
Sku supports Vitest as an alternate to Jest since v15

**We recommend using vitest with the vite bundler and in an ESM project**, however these are not requirements.

## Migrating to Vitest

Migrating to Vitest should be fairly straightforward.
Before making any changes, it is highly recommend to read [the Vitest documentation](https://vitest.dev/guide/) to familiarise yourself with the API and features, as well as to better understand the differences between Jest and Vitest.

To enable Vitest in `sku`, first install the required dependencies:

```bash
pnpm add -D vitest @sku-lib/vitest
```

Then, configure [`testRunner`][test runner] in your `sku` config:

```typescript
// sku.config.ts
import type { SkuConfig } from 'sku';

export default {
  testRunner: 'vitest',
  ...
} satisfies SkuConfig;
```

`sku` will now invoke [the `vitest` CLI][Vitest CLI] instead of the `jest` CLI when running `sku test`.

#### Key differences between Vitest and Jest

Vitest has strong compatibility with the Jest API, however it still has differences that may affect your tests.
Differences that affect sku projects are listed here.
The full list is documented in the Vitest documentation: [Migrating from Jest to Vitest].

To automate most of the migration process, a codemod is available.
Note that additional changes may still be required after running this codemod.

```sh
pnpm dlx @sku-lib/codemod jest-to-vitest .
```

**Default Watch Mode**

`vitest` defaults to watch mode when running tests.
To run tests without watch mode you can use the `--run` flag:

```json
// package.json
{
  "scripts": {
    // ...
    "test": "sku test --run"
  }
}
```

Watch mode won't trigger in CI environments, so it's safe to omit the flag in your pipeline.

**Code Coverage Dependencies**

Vitest does not install code coverage dependencies by default.
To collect code coverage, pass the `--coverage` flag to the `sku test` command and install the `@vitest/coverage-v8` package:

```bash
pnpm add -D @vitest/coverage-v8
```

**Testing Library Matchers**

If your test setup file includes an import for `@testing-library/jest-dom`, you may need to change this to `@testing-library/jest-dom/vitest`:

```diff
// test-setup.ts
- import '@testing-library/jest-dom';
+ import '@testing-library/jest-dom/vitest';
```

**Globals Disabled**

Jest enables global APIs such as `it`, `describe`, `beforeAll`, etc., by default.
Vitest does not.
Sku adopts the Vitest defaults, meaning you will need to explicitly import these test functions.
This change results in a cleaner global namespace, better type safety, clearer dependencies and is more consistency with modern javascript practices.

```diff
+ import { describe, expect, it } from 'vitest';
```

Be aware that since globals are disabled, some common libraries like `testing-library` will not run auto DOM cleanup.
If using these libraries, you will need to add cleanup to your configured `setupTests` file.

```diff
import '@testing-library/jest-dom/vitest';

+ import { cleanup } from '@testing-library/react';
+ import { afterEach } from 'vitest';

+ afterEach(cleanup)
```

[Vitest]: https://vitest.dev/
[Jest's current limitations with ESM]: https://jestjs.io/docs/ecmascript-modules
[test runner]: ./docs/configuration.md#testrunner
[Vitest CLI]: https://vitest.dev/guide/cli.html
[codemod]: https://codemod.com/registry/jest-vitest
[Migrating from Jest to Vitest]: https://vitest.dev/guide/migration.html#jest
[Vitest globals]: https://vitest.dev/config/#globals
