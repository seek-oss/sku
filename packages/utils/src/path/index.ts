export { toPosixPath } from './toPosixPath.js';
