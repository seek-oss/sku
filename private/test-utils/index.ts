export { ListExternalsWebpackPlugin } from './ListExternalsWebpackPlugin.ts';
export { dirContentsToObject } from './dirContentsToObject.ts';
export { makeStableHashes, makeStableViteHashes } from './skuConfig.ts';
export { getStoryPage, getTextContentFromFrameOrPage } from './storybook.ts';
export { getPort } from './getPort.ts';
