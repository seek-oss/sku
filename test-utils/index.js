// @ts-check
export { ListExternalsWebpackPlugin } from './ListExternalsWebpackPlugin.js';
export { getAppSnapshot } from './appSnapshot.js';
export { startAssetServer } from './assetServer.js';
export { dirContentsToObject } from './dirContentsToObject.js';
export { run, runSkuScriptInDir } from './process.ts';
export { makeStableHashes } from './skuConfig.js';
export { getStoryPage, getTextContentFromFrameOrPage } from './storybook.js';
export { waitForUrls } from './waitForUrls.js';
