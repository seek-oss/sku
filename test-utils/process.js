const { promisify } = require('util');
// spwan can't be promisify'ed so we use execFile, which just wraps spawn and can be promisify'ed
// https://github.com/nodejs/node/blob/2f369ccacfb60c034de806f24164524910301825/lib/child_process.js#L326
const execFile = promisify(require('child_process').execFile);
const gracefulSpawn = require('sku/lib/gracefulSpawn');

const skuBin = require.resolve('sku/bin/sku.js');

/**
 * @param {string} file The name or path of the executable file to run
 * @param {string[]} [args] List of string arguments
 * @param {import('child_process').ExecFileOptions} [options] Process execution options
 * @return {Promise<{ stdout: string; stderr: string; child: import('child_process').ChildProcess }>}
 */
const run = async (file, args = [], options = {}) => {
  try {
    const promise = execFile(file, args, {
      env: process.env,
      ...options,
    });
    // Yes, this is a bit odd.
    // If we don't spread the promise the child process is not returned, only stdout and stderr.
    const result = await promise;
    return { ...promise, ...result };
  } catch (error) {
    // print the stdout of a failed command so we can see it in the Jest output
    if (error.stdout) {
      console.warn(error.stdout);
    }
    if (error.stderr) {
      console.warn(error.stderr);
    }
    throw error;
  }
};

/**
 * @param {string} script The sku script to run
 * @param {string} cwd The current working directory of the child process
 * @param {string[]} [args] List of string arguments
 * @param {import('child_process').ExecFileOptions | import('child_process').SpawnOptions} [options] Process execution options
 */
const runSkuScriptInDir = async (script, cwd, args = [], options = {}) => {
  const processOptions = {
    cwd,
    env: {
      ...process.env,
      STORYBOOK_DISABLE_TELEMETRY: 1,
    },
    // Increased from 1024 * 1024 because Storybook can produce very large outputs.
    // https://nodejs.org/docs/latest-v18.x/api/child_process.html#child_process_child_process_exec_command_options_callback
    maxBuffer: 5 * 1024 * 1024,
    ...options,
  };

  // When starting a dev server, return a hook to the running process
  if (/^(start|storybook|serve)/.test(script)) {
    return gracefulSpawn(skuBin, [script, ...args], {
      stdio: 'inherit',
      ...processOptions,
    });
  }

  // Otherwise, resolve the promise when the script finishes
  return run(skuBin, [script, ...args], processOptions);
};

module.exports = {
  run,
  runSkuScriptInDir,
};
