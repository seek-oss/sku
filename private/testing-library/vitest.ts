// Re-export vitest setup from cli-testing-library
import 'cli-testing-library/vitest';
