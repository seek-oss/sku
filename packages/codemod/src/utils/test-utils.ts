import { readFile } from 'node:fs/promises';
import path from 'node:path';
import { expect } from 'vitest';
import { isValidTransform, type Transform } from './types.js';

type FileInfo = {
  source: string;
  path: string;
};

type InlineTestProps = {
  transform: Transform;
  input: FileInfo;
  expectedOutput: string;
};

type TestProps = {
  dirName: string;
  transformName: string;
  fixtureName?: string;
  testOptions?: { extension?: string };
};

const runInlineTest = async ({
  transform,
  input,
  expectedOutput,
}: InlineTestProps) => {
  if (!isValidTransform(transform)) {
    throw new Error('Invalid transform function provided');
  }

  const output = transform(input.source);
  if (!output) {
    throw new Error('Transform failed');
  }

  expect(output.trim()).toEqual(expectedOutput.trim());

  return output;
};

export const runTest = async ({
  dirName,
  transformName,
  fixtureName,
  testOptions = {},
}: TestProps) => {
  const testFixtureName = fixtureName || transformName;

  const fixtureDir = path.join(dirName, '..', '__testfixtures__');
  const inputPath = path.join(
    fixtureDir,
    `${testFixtureName}.input.${testOptions.extension}`,
  );

  // Assumes transform is one level up from __tests__ directory
  const transformImport = import(path.join(dirName, '..', transformName));
  const sourceRead = readFile(inputPath, 'utf8');
  const expectedOutputRead = readFile(
    path.join(fixtureDir, `${testFixtureName}.output.${testOptions.extension}`),
    'utf8',
  );

  const [{ transform }, source, expectedOutput] = await Promise.all([
    transformImport,
    sourceRead,
    expectedOutputRead,
  ]);

  return runInlineTest({
    transform,
    input: {
      path: inputPath,
      source,
    },
    expectedOutput,
  });
};

export const runNoChangeTest = async ({
  dirName,
  transformName,
  fixtureName,
  testOptions = {},
}: TestProps) => {
  const inputPath = path.join(
    dirName,
    '..',
    '__testfixtures__',
    `${fixtureName}.unchanged.${testOptions.extension}`,
  );

  const sourceRead = readFile(inputPath, 'utf8');
  const transformImport = import(path.join(dirName, '..', transformName));

  const [{ transform }, source] = await Promise.all([
    transformImport,
    sourceRead,
  ]);
  const output = await transform(source);

  expect(output).toEqual(null);
};
