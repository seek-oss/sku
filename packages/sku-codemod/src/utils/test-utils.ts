import { readFile } from 'node:fs/promises';
import path from 'node:path';
import { expect } from 'vitest';

type FileInfo = {
  source: string;
  path: string;
};

const runInlineTest = async (
  transform: (source: string) => string | false,
  input: FileInfo,
  expectedOutput: string,
) => {
  const output = await transform(input.source);
  if (!output) {
    throw new Error('Transform failed');
  }
  const expectation = (o: string) =>
    expect(o.trim()).toEqual(expectedOutput.trim());
  expectation(output);
  return output;
};

export const runTest = async (
  dirName: string,
  transformName: string,
  testFilePrefix?: string,
  testOptions: Record<string, unknown> = {},
) => {
  let localTestFilePrefix = testFilePrefix;
  if (!testFilePrefix) {
    localTestFilePrefix = transformName;
  }

  // Assumes transform is one level up from __tests__ directory
  const transform = (await import(path.join(dirName, '..', transformName)))
    .transform;
  const fixtureDir = path.join(dirName, '..', '__testfixtures__');
  const inputPath = path.join(
    fixtureDir,
    `${localTestFilePrefix}.input.${testOptions.extension}`,
  );
  const source = await readFile(inputPath, 'utf8');
  const expectedOutput = await readFile(
    path.join(
      fixtureDir,
      `${localTestFilePrefix}.output.${testOptions.extension}`,
    ),
    'utf8',
  );
  return runInlineTest(
    transform,
    {
      path: inputPath,
      source,
    },
    expectedOutput,
  );
};

export const runNoChangeTest = async (
  dirName: string,
  transformName: string,
  testFilePrefix: string,
  testOptions: Record<string, unknown>,
) => {
  const transform = (await import(path.join(dirName, '..', transformName)))
    .transform;
  const inputPath = path.join(
    dirName,
    '..',
    '__testfixtures__',
    `${testFilePrefix}.unchanged.${testOptions.extension}`,
  );
  const source = await readFile(inputPath, 'utf8');
  const output = await transform(source);
  expect(output).toEqual(false);
};
