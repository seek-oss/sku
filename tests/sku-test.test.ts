import { describe, it } from 'vitest';
import path from 'node:path';
import { runSkuScriptInDir } from '@sku-private/test-utils';
import { createRequire } from 'node:module';
import { cwd } from 'node:process';
import { stripVTControlCharacters } from 'node:util';

const require = createRequire(import.meta.url);

const appDir = path.dirname(
  require.resolve('@sku-fixtures/sku-test/sku.config.ts'),
);

describe.for(['vitest', 'jest'])('[%s]: sku-test', (testRunner) => {
  const args = testRunner === 'vitest' ? ['--config=sku.config.vitest.ts'] : [];
  it('should run tests', async ({ expect }) => {
    await expect(
      runSkuScriptInDir('test', appDir, {
        args,
      }),
    ).resolves.not.toThrowError();
  });

  it(`should pass through unknown flags to ${testRunner}`, async ({
    expect,
  }) => {
    const child = await runSkuScriptInDir('test', appDir, {
      args: [...args, '--coverage'],
    });
    const output = (child?.stdout as string).replaceAll(cwd(), 'sku');

    expect(
      stripVTControlCharacters(
        // strip out the dynamic parts of the output
        output.replaceAll(/Start at .*|Duration .*/g, ''),
      ),
    ).toMatchSnapshot();
  });
});
