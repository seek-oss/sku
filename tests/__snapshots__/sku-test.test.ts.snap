// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sku-test should pass through unknown flags to jest 1`] = `
"sku/fixtures/sku-test/src/index.test.ts
"
`;
